### PR TITLE
Add Custom Build for ASM files

### PR DESCRIPTION
**Pull request by Brett208**
When coupled with NetFixClient, produces the following 2 errors:
```
Error	LNK1120	1 unresolved externals	NetFixClient	\NetFixClient\Release\NetFixClient.dll	1

Error	LNK2019	unresolved external symbol "public: void __thiscall OP2Internal::TApp::NetShutdown(int)" (?NetShutdown@TApp@OP2Internal@@QAEXH@Z) referenced in function "private: void __thiscall OPUNetGameSelectWnd::CleanupGurManager(void)" (?CleanupGurManager@OPUNetGameSelectWnd@@AAEXXZ)	NetFixClient	\NetFixClient\OPUNetGameSelectWnd.obj	1
```
I think the custom build script is actually compiling a .obj file?

### CODE REVIEW

**Comment by Brett208** [comment]
Updated to most recent version of branch. On release builds, also getting the following warning which may be relevant:

Warning	LNK4042	object specified more than once; extras ignored	OP2Internal	\NetFixClient\OP2Internal\OP2Internal.obj	1

File: OP2Internal.vcxproj [hunk 1]
```diff
@@ -45,12 +45,34 @@
...
+    <CustomBuild>
+      <Command>"C:\Program Files (x86)\NASM\nasm" -fwin32 "%(FullPath)" -o %(Filename).obj</Command>
+    </CustomBuild>
+    <CustomBuild>
+      <Message>Build ASM File</Message>
+    </CustomBuild>
+    <CustomBuild>
+      <TreatOutputAsContent>true</TreatOutputAsContent>
+      <Outputs>%(Filename).obj</Outputs>
+    </CustomBuild>
```
Comment (DanRStevens) [inline OP2Internal.vcxproj]: Hmm, why are these split among separate `CustomBuild` tags? I would assume the `Command`, `Message`, `Outputs`, and such should all be under the same tag.

File: OP2Internal.vcxproj [hunk 2]
```diff
@@ -45,12 +45,34 @@
...
+  </ItemDefinitionGroup>
+  <ItemDefinitionGroup Condition="'$(Configuration)|$(Platform)'=='Release|Win32'">
+    <CustomBuild>
+      <Command>"C:\Program Files (x86)\NASM\nasm" -fwin32 "%(FullPath)" -o %(Filename).obj</Command>
+      <Message>Build ASM File</Message>
+      <TreatOutputAsContent>true</TreatOutputAsContent>
+      <Outputs>%(Filename).obj</Outputs>
+    </CustomBuild>
```
Comment (DanRStevens) [inline OP2Internal.vcxproj:55]: This looks like a duplicate of the above, but all under a single `CustomBuild` tag, like I expected.

File: OP2Internal.vcxproj [hunk 3]
```diff
@@ -45,12 +45,34 @@
...
   </ItemDefinitionGroup>
   <ItemGroup>
     <ClCompile Include="DummyMain.cpp" />
   </ItemGroup>
   <ItemGroup>
-    <None Include="OP2Internal.asm" />
+    <CustomBuild Include="OP2Internal.asm">
+      <ExcludedFromBuild Condition="'$(Configuration)|$(Platform)'=='Release|Win32'">false</ExcludedFromBuild>
+      <FileType>Document</FileType>
+      <Outputs Condition="'$(Configuration)|$(Platform)'=='Release|Win32'">OP2Internal.obj;%(Outputs)</Outputs>
```
Comment (DanRStevens) [inline OP2Internal.vcxproj]: The semicolon separated list here looks odd. There should be a single output file.

Additionally the tagging of only "Release" doesn't seem right. Is it only compiling the assembly file in release mode?

**Review by DanRStevens — COMMENTED** [review]
Ok, thank you very much for getting this done. I noticed a few odd looking points in the project configuration, though it sounds like you've basically got the .asm file to compile into a .obj file using NASM.

I'm wondering if the error might be from specifying a list of output files, rather than a single output file. The settings seem to give both an explicit filename, and a variable which expands to the same filename.

I also suspect it might only be building the .obj for release builds, and not for debug builds.

**Comment by DanRStevens** [comment]
While taking a second look at this, I think we should avoid the hardcoded path to NASM, and perhaps instead rely on NASM being installed in the PATH. There are separate default install paths for x86 and x64 versions of NASM. It may be wise to not rely on either.

In particular, this should change:
```
<Command>"C:\Program Files (x86)\NASM\nasm" -fwin32 "%(FullPath)" -o %(Filename).obj</Command>
```

In the above command, the path is hardcoded for an x86 install. It will fail with the default x64 install path of `%ProgramFiles%\NASM\` = `C:\Program Files\NASM`.

----

With the path change, removal of duplicate settings, and ensuring it builds for both Debug and Release, I think we'll be able to close this one off. PR #12 should take care of installing NASM on AppVeyor.

File: OP2Internal.vcxproj (hunk 1 above)
Comment (Brett208) [reply OP2Internal.vcxproj]: EDIT: It looks like you can just access the custom build tool properties from either clicking on the file or opening the project page.

File: OP2Internal.vcxproj (hunk 3 above)
Comment (Brett208) [reply OP2Internal.vcxproj]: I messed up at some point and was setting the property under release and debug only at one point.

**Comment by Brett208** [comment]
Latest push fixes the LNK4042 on the release build. It also cleans up the XML in the project file.

Unfortunately, the 2 errors above still exist. I'm not sure where to proceed from here? It appears to be building the asm file into an obj file and the linker seems to be finding it. At least I'm not getting any errors about the process. Is it a problem that we are naming the obj file as the same name as the project file?

-Brett

**Comment by DanRStevens** [comment]
Before I forget, I was curious was this setting was for:
```
<TreatOutputAsContent>true</TreatOutputAsContent>
```

I haven't been able to find any documentation on it.

**Comment by Brett208** [comment]
As all I know is the tooltip says "specify whether the output file is a deployable content."

**Comment by DanRStevens** [comment]
Ahh, that's probably for resource files and such that get copied to the output folder (rather than the intermediate folder), or packaged with the exe. That might include game resources, such as graphics, sounds, or music, or a readme file.

As NASM produces a .obj file which is later used by the linker to build the exe itself, I'd say no here, it's not deployable content.

**Comment by DanRStevens** [comment]
I noticed a slight discrepancy in the name mangling between the linker error message and what was defined in the .asm file. I've corrected the .asm file to match.

Possible reasons why:
- A newer compiler version updated its name mangling rules (I somehow doubt this, as it would break compatibility with older C++ ABI versions)
- The method signature was updated
  - in revision 670a4044 a return type of `void` was added. That might affect the name mangling, though I believe the old code was incorrect and might not have even compiled, hence it wouldn't have had a mangled name.
   - the parameter is of type `int`, but is named `bSendQuitMessage`, suggesting it is a bool. Perhaps the parameter type was updated from `bool` to `int`, and that changed the name mangling.
- It was always wrong in the repository (but perhaps corrected and never committed during previous NetFixClient releases)

As for the update, the newer name mangling now more closely matches the corresponding export from Outpost2.exe, as verified by dumping the symbols from `Outpost2.lib` from the Outpost2DLL project:
```
nm Outpost2DLL/Lib/Outpost2DLL.lib | grep NetShutdown
00000000 I __imp_?NetShutdown@TApp@@QAEXH@Z
00000000 T ?NetShutdown@TApp@@QAEXH@Z
```

Notice the `QAEXH`, and compare to the recent fix from `QAEHH`:
```
-Export 0x00487E30, ?NetShutdown@TApp@OP2Internal@@QAEHH@Z
+Export 0x00487E30, ?NetShutdown@TApp@OP2Internal@@QAEXH@Z
```

----

I think this is good to merge now.

**Comment by Brett208** [comment]
The immediate error was fixed. It uncovered about 42 new errors that I will push into a  separate issue.

-Brett

**Comment by DanRStevens** [comment]
Excellent.

Tagging Issue #9.

**Pull request merged**